### ISSUE TITLE
Startup error: undefined method `to_h' for #<Array

### ISSUE THREAD

**Issue by ginjo** [boazsegev/plezi#7]
I'm following the readme instructions, however Plezi won't startup on my system (SnowLeopard, ruby 1.9.3v551). Here's my gem list & stack trace. Happy to provide more... Thanks.

```
MacBook:plezi_local_demo wbr$ bundle list
Gems included by the bundle:
  * bundler (1.7.11)
  * greactor (0.1.0)
  * grhttp (0.1.1)
  * plezi (0.11.2)
  * rake (10.4.2)
MacBook:plezi_local_demo wbr$ bundle exec plezi s
/Users/wbr/Desktop/plezi_local_demo/vendor/bundle/ruby/1.9.1/gems/grhttp-0.1.1/lib/grhttp/hpack.rb:212: warning: already initialized constant STATIC_LIST
/Users/wbr/Desktop/plezi_local_demo/vendor/bundle/ruby/1.9.1/gems/grhttp-0.1.1/lib/grhttp/hpack.rb:274: warning: already initialized constant STATIC_LENGTH
/Users/wbr/Desktop/plezi_local_demo/vendor/bundle/ruby/1.9.1/gems/grhttp-0.1.1/lib/grhttp/hpack.rb:277:in `<class:HPACK>': undefined method `to_h' for #<Array:0x00000100b4d6f0> (NoMethodError)
        from /Users/wbr/Desktop/plezi_local_demo/vendor/bundle/ruby/1.9.1/gems/grhttp-0.1.1/lib/grhttp/hpack.rb:6:in `<class:HTTP2>'
        from /Users/wbr/Desktop/plezi_local_demo/vendor/bundle/ruby/1.9.1/gems/grhttp-0.1.1/lib/grhttp/hpack.rb:5:in `<module:Base>'
        from /Users/wbr/Desktop/plezi_local_demo/vendor/bundle/ruby/1.9.1/gems/grhttp-0.1.1/lib/grhttp/hpack.rb:4:in `<module:GRHttp>'
        from /Users/wbr/Desktop/plezi_local_demo/vendor/bundle/ruby/1.9.1/gems/grhttp-0.1.1/lib/grhttp/hpack.rb:3:in `<top (required)>'
        from /Users/wbr/Desktop/plezi_local_demo/vendor/bundle/ruby/1.9.1/gems/grhttp-0.1.1/lib/grhttp.rb:18:in `require'
        from /Users/wbr/Desktop/plezi_local_demo/vendor/bundle/ruby/1.9.1/gems/grhttp-0.1.1/lib/grhttp.rb:18:in `<top (required)>'
        from /Users/wbr/Desktop/plezi_local_demo/vendor/bundle/ruby/1.9.1/gems/plezi-0.11.2/lib/plezi.rb:15:in `require'
        from /Users/wbr/Desktop/plezi_local_demo/vendor/bundle/ruby/1.9.1/gems/plezi-0.11.2/lib/plezi.rb:15:in `<top (required)>'
        from /usr/local/rvm/gems/ruby-1.9.3-p551@global/gems/bundler-1.7.11/lib/bundler/runtime.rb:76:in `require'
        from /usr/local/rvm/gems/ruby-1.9.3-p551@global/gems/bundler-1.7.11/lib/bundler/runtime.rb:76:in `block (2 levels) in require'
        from /usr/local/rvm/gems/ruby-1.9.3-p551@global/gems/bundler-1.7.11/lib/bundler/runtime.rb:72:in `each'
        from /usr/local/rvm/gems/ruby-1.9.3-p551@global/gems/bundler-1.7.11/lib/bundler/runtime.rb:72:in `block in require'
        from /usr/local/rvm/gems/ruby-1.9.3-p551@global/gems/bundler-1.7.11/lib/bundler/runtime.rb:61:in `each'
        from /usr/local/rvm/gems/ruby-1.9.3-p551@global/gems/bundler-1.7.11/lib/bundler/runtime.rb:61:in `require'
        from /usr/local/rvm/gems/ruby-1.9.3-p551@global/gems/bundler-1.7.11/lib/bundler.rb:134:in `require'
        from /Users/wbr/Desktop/plezi_local_demo/environment.rb:21:in `<top (required)>'
        from /Users/wbr/Desktop/plezi_local_demo/plezi_local_demo:5:in `load'
        from /Users/wbr/Desktop/plezi_local_demo/plezi_local_demo:5:in `<top (required)>'
        from /Users/wbr/Desktop/plezi_local_demo/vendor/bundle/ruby/1.9.1/gems/plezi-0.11.2/bin/plezi:71:in `load'
        from /Users/wbr/Desktop/plezi_local_demo/vendor/bundle/ruby/1.9.1/gems/plezi-0.11.2/bin/plezi:71:in `rescue in <top (required)>'
        from /Users/wbr/Desktop/plezi_local_demo/vendor/bundle/ruby/1.9.1/gems/plezi-0.11.2/bin/plezi:71:in `<top (required)>'
        from /Users/wbr/Desktop/plezi_local_demo/vendor/bundle/ruby/1.9.1/bin/plezi:23:in `load'
        from /Users/wbr/Desktop/plezi_local_demo/vendor/bundle/ruby/1.9.1/bin/plezi:23:in `<main>'

```

**Comment by boazsegev** [comment]
Hi William, thank you for reporting the issue.

This was actually part of GRHttp's code base for future HTTP/2 support, which isn't really implemented just yet.

If you want to fix the issue on your computer, you can update the `hpack.rb` file in the `grhttp` gem so that line 277 will read:

```ruby
"111111111111111111111111111111"].each_with_index.with_object({}) {|a, h| h[a[0]] = a[1] }
```

I have rewritten the code to avoid the `Array#to_h` method that was introduced only with Ruby version 2.1.0... `Array#to_h` is no longer in use within the Plezi project (including Greater and GRHttp).

But...

Plezi was written with the Ruby 2.2.0 family in mind - including performance optimizations. It should work mostly with no issues on older systems, but that's not really something I was worried about. From my experience, it's quite easy to have multiple versions of Ruby on the same system and I believe that new projects would benefit from starting with the most updated Ruby version...

My recommendation to you is to install Ruby 2.2.3 in addition to your existing version(s) and use Plezi with the latest Ruby version.

Again, thank you both for your interest in Plezi and for reporting this issue. It's people like you that make the world of open source so amazing.

P.S.

I'm now working on Plezi version 0.12.0 (almost done with this iteration), including an updated Reactor (v.0.2.0 resolves some SSL issues, includes minor optimizations and has some added functionality that GRHttp 0.2.0 leverages), an updated HTTP server (GRHttp is closer to supporting HTTP/2 and enjoys some fine-tuning) and some new requested features, such as a scalable (only called once) unicast failure callback that allows a fallback routine, such as push notification or message persistent storage.

Because of this work, it might take me a bit before I release the fixed version... If you're in a hurry, you can either install the edge version for all three gems directly from github or manually edit the `hpack.rb` file.

Bo.

**Comment by ginjo** [comment]
Thanks @boazsegev  for the informative reply. I thought it might be something like that. I'll try the patch, or the newer Ruby (gotta free up some HD space), and see how it goes. Looking forward to trying it out!

Bill

**Comment by boazsegev** [comment]
Hi Bill,

I released the updated Plezi version (0.12.0), after performing a big amount of recoding on the server (with still more to come, but this is a never-ending process)...

Please let me know how it works out for you and how I can improve Plezi for common use-cases.

Thanks!
